hw_diag/utilities/network_watchdog.py

```diff
@@ -0,0 +1,146 @@
+
+        # Set up logger
+        self.logger = get_logger(__name__)
+        # Add rotating log handler to the logger
+        handler = RotatingFileHandler(self.log_file_path, maxBytes=self.MAX_LOG_SIZE, backupCount=3)
+        handler.setLevel(logging.INFO)
+        handler.setFormatter(
+            logging.Formatter("%(asctime)s - [%(levelname)s]:(%(lineno)d) - %(message)s"))
+        self.logger.addHandler(handler)
+
+        self.systemd_proxy = Systemd()
+        self.network_manager_unit = self.systemd_proxy.get_unit(DBusIds.NETWORK_MANAGER_UNIT_NAME)
+        self.network_manager = NetworkManager()
+
+        # Log the watchdog intro
+        self.logger.info("Watchdog is started!")
+
+    def __del__(self):
+        if hasattr(self, 'temp_dir'):
+            self.temp_dir.cleanup()
+
+    def have_internet(self, timeout=10) -> bool:
+        try:
+            socket.setdefaulttimeout(timeout)
+            sock = socket.create_connection((self.PUBLIC_DNS_SERVER, self.PUBLIC_DNS_PORT))
+            sock.close()
+            return True
+        except Exception as e:
+            self.logger.info(f"internet not accessible: {e}")
+            return False
+
+    def is_connected(self) -> bool:
+        self.logger.info("Checking the network connectivity.")
+
+        # Log more details about the network connectivity and internet connectivity
+        self.logger.info(f"Network manager state: {self.network_manager.get_connect_state()}")
+        self.logger.info(f"Internet connectivity: {self.have_internet()}")
+
+        nm_connected = self.network_manager.is_connected()
+        return nm_connected
+
+    def restart_network_manager(self):
+        """Restart hostOS NetworkManager service"""
+        nm_restarted = self.network_manager_unit.wait_restart()
+        self.logger.info(f"Network manager restarted: {nm_restarted}")
+
+    def get_last_restart(self) -> datetime:
+        try:
+            last_restart = KeyStore(self.state_file_path).get(self.LAST_RESTART_KEY)
+            return datetime.strptime(last_restart, self.LAST_RESTART_DATE_FORMAT)
+        except Exception as e:
+            self.logger.info(f"Can not find the previous restart time: {e}")
+            return datetime.min
+
+    def save_last_restart(self) -> None:
+        store = KeyStore(self.state_file_path)
+        store.set(self.LAST_RESTART_KEY, datetime.now().strftime(self.LAST_RESTART_DATE_FORMAT))
+        self.logger.info("Saved the last_restart.")
+
+    def ensure_network_connection(self) -> None:
+        self.logger.info("Running the watchdog...")
+
+        if self.is_connected():
+            self.lost_count = 0
+            self.logger.info("Network is working.")
+        else:
+            self.lost_count += 1
+            self.logger.warning(
+                f"Network is not connected! Lost connectivity count={self.lost_count}")
+
+            if self.lost_count > self.NM_RESTART_THRESHOLD:
```

Comment:
I'm allergic to deeply nested code. Better if broken into smaller methods.

Reply:
Fair enough. I refactored to avoid the deeply nested codes.